### DISCUSSION TITLE
Error when setting up Docker container

### DISCUSSION THREAD

**Discussion by timmarkhuff** [Robotic-Decision-Making-Lab/blue#118] [Q&A] [answered]
I am struggling to set up the docker container.

I run `sudo docker build -t blue .` and I get:
ERROR: Could not open requirements file: [Errno 2] No such file or directory: '/root/ws_blue/src/blue/requirements-build.txt'

Are there some instructions on how to set up the docker container?

Also, why does the Dockerfile reference a requirements.txt file? Shouldn't it be self-contained and not rely on external requirements files?

**Comment by evan-palmer** [comment]
Thanks for checking the project out @timmarkhuff! I apologize for the current lack of documentation. I unfortunately haven't had time yet to setup documentation for the project. This project is only a couple months old now. Adding documentation is in my immediate pipeline though (see #31). I do my best to respond quickly here in the discussions, so if you run into any other problems feel free to send me a message or to post a discussion. 

The easiest way (what I recommend) to use the Docker environment is to use the [devcontainer](https://github.com/Robotic-Decision-Making-Lab/blue/tree/main/.devcontainer) configuration with VSCode ([here](https://code.visualstudio.com/docs/devcontainers/containers) is the link to the devcontainer documentation). This is the shortest path to working with the project as the devcontainer configuration manages the build context for you (the reason for the error that you have mentioned) and has the run arguments and other environment variables already set up for you to enable running Gazebo and the other visualization tools like RViz2. If you are building this on a laptop, I recommend modifying the `colcon build` command on [line 177](https://github.com/Robotic-Decision-Making-Lab/blue/blob/7b83a99157383465f51dff860b676f5ec3d70b0d/.docker/Dockerfile#L177C10-L177C10) to be:

```MAKEFLAGS="-j1 -l1" colcon build```

You can configure the number of threads as you see fit, but this is the safest solution. The project has to build `ros_gz` from source (*shakes fist*) and colcon is not the most friendly build system for RAM. I'm planning to build a more user-friendly image on the GitHub servers eventually, but haven't had the time to do that yet.

To answer your last question regarding why the Dockerfile references a requirements file, this is to help improve the maintainability of the project. Instead of managing dependencies in both the Dockerfile and a requirements file (for users that do not want to install the system using Docker), the Dockerfile just uses the requirements file. You will find that the Dockerfile also uses rosdep and vcs to manage other dependencies (hence also using the `.repos` file and `package.xml` files).

**Comment by evan-palmer** [comment, marked as answer]
So I have been meaning to cleanup the Dockerfile for a bit now and this thread has given me the motivation to update that. I've just pushed some new changes to the branch `120-feature-cleanup-dockerfile` that improves support for running the project in Docker without the devcontainer configuration. There isn't any need to build the image either as I've added a new pipeline to build images and save images on the GitHub container registry for use with either NVIDIA driver support or use with other drivers (see the project images [here](https://github.com/Robotic-Decision-Making-Lab/blue/pkgs/container/blue)). I'm hoping to get the PR merged sometime tomorrow, but that is subjective to how well my convex optimization homework is going :grimacing:. In the meantime, you can pull the latest changes and run the following for use with systems that don't have NVIDIA driver support

```bash
docker compose -f path/to/blue/.docker/compose/nouveau-desktop.yaml up -d
```
Then start a new interactive container with

```bash
docker exec -it <name-of-container> /bin/bash
```
If your platform supports the NVIDIA drivers, you can perform the same commands mentioned above, but instead use the `nvidia-compose.yaml` compose script. Once you are finished, you can run 
```bash
docker compose -f path/to/blue/.docker/compose/nouveau-desktop.yaml down
```
to stop the container. At the time of writing this post, I haven't fixed #25, so the Humble build isn't ready yet, but Iron and Rolling both work. I also still need to test the Nouveau drivers out on my laptop, but the NVIDIA container works on my desktop and the Nouveau version has worked in the past. Let me know if this solves some of your problems @timmarkhuff 

EDIT: I updated the compose script name based off of the PR

**Reply by evan-palmer** [reply]
Note that you will need to sign into the Docker container registry to download the images

**Comment by timmarkhuff** [comment]
@evan-palmer , thank you so much! I will give this a try. 

I am on a team at the University of Washington and we are using Blue ROV 2 with ROS for our research, so we were very pleased to find your package. We hope to give you feedback on it and contribute to the project if possible.

**Comment by evan-palmer** [comment]
Very cool! Do you happen to work with Aaron Marburg at APL? If not, you may find him to be a valuable resource. I know that he's worked a bit with the BlueROV in the past

**Comment by amarburg** [comment]
Yes!  They are a student team I'm advising....

**Comment by timmarkhuff** [comment]
Hi @evan-palmer , we got our Blue ROV 2 set up, and we're diving back into this ROS 2 package now. I've hit another snag with the docker container, and I'm hoping you can help.

I don't see a `120-feature-cleanup-dockerfile` branch so I am assuming it has been merged in and deleted, and I have been working in the `main` branch.

I am getting a 404 , when I try to access this link you posted: https://github.com/Robotic-Decision-Making-Lab/blue/pkgs/container/blue

I followed your instructions and logged into GHCR with Docker: `docker login ghcr.io -u YOUR_GITHUB_USERNAME -p YOUR_GENERATED_TOKEN`

Then I tried running docker compose, but it seems I don't have access: 
```
PS C:\Users\timma\blue\.docker\compose> docker-compose -f nouveau-desktop.yaml up -d
time="2023-10-18T11:31:53-07:00" level=warning msg="The \"DISPLAY\" variable is not set. Defaulting to a blank string."
time="2023-10-18T11:31:53-07:00" level=warning msg="The \"XDG_RUNTIME_DIR\" variable is not set. Defaulting to a blank string."
time="2023-10-18T11:31:53-07:00" level=warning msg="The \"WAYLAND_DISPLAY\" variable is not set. Defaulting to a blank string."
time="2023-10-18T11:31:53-07:00" level=warning msg="The \"PULSE_SERVER\" variable is not set. Defaulting to a blank string."
[+] Running 1/1
 ✘ blue Error                                                                                                                                                                 0.8s
Error response from daemon: denied
```

Do you have any idea what might be going wrong?

**Comment by evan-palmer** [comment]
My bad! The package should be public now. I didn’t realize that package visibility defaulted to private on GitHub
